src/3D/Camera.cpp

```diff
@@ -9,28 +9,97 @@
 
 #include "Camera.h"
 #include "3D/LandIsland.h"
+#include "ECS/Components/Transform.h"
 #include "ECS/Registry.h"
 #include "Game.h"
 
 #include <glm/gtx/euler_angles.hpp>
+#include <glm/gtx/intersect.hpp>
+#include <glm/gtx/norm.hpp>
+
+#define OUT
 
 using namespace openblack;
 
-glm::mat4 Camera::getRotationMatrix() const
+Camera::Camera(glm::vec3 position, glm::vec3 rotation)
+    : _position(position)
+    , _rotation(glm::radians(rotation))
+    , _dv(0.0f, 0.0f, 0.0f)
+    , _dwv(0.0f, 0.0f, 0.0f)
+    , _dsv(0.0f, 0.0f, 0.0f)
+    , _ddv(0.0f, 0.0f, 0.0f)
+    , _duv(0.0f, 0.0f, 0.0f)
+    , _drv(0.0f, 0.0f, 0.0f)
+    , _projectionMatrix(1.0f)
+    , _velocity(0.0f, 0.0f, 0.0f)
+    , _hVelocity(0.0f, 0.0f, 0.0f)
+    , _rotVelocity(0.0f, 0.0f, 0.0f)
+    , _accelFactor(0.001f)
+    , _movementSpeed(4.0f)
+    , _maxMovementSpeed(0.005f)
+    , _maxRotationSpeed(0.005f)
+    , _mouseIsDown(false)
+    , _mouseIsMoving(false)
+    , _shiftHeld(false)
+{
+}
+
+glm::mat4 Camera::GetRotationMatrix() const
 {
 	return glm::eulerAngleZXY(_rotation.z, _rotation.x, _rotation.y);
 }
 
 glm::mat4 Camera::GetViewMatrix() const
 {
-	return getRotationMatrix() * glm::translate(glm::mat4(1.0f), -_position);
+	return GetRotationMatrix() * glm::translate(glm::mat4(1.0f), -_position);
 }
 
 glm::mat4 Camera::GetViewProjectionMatrix() const
 {
 	return GetProjectionMatrix() * GetViewMatrix();
 }
 
+bool Camera::RaycastCamToLand(OUT glm::vec3* hit)
```

Comment:
like this, just looks good imo

Reply:
We don't use this style anywhere else and seems kind of arbitrary to start doing it now just here.
Not to mention it's not a good place to add global defines.
`OUT` is also a very 1980s indicator in c++. Might be some modern compiler hints that do this.

Reply:
On another note, it might be better to have that parameter be a reference. This would add a linguistic guarantee that there is valid data there.
With a naked pointer, there are more chances of runtime errors.

Reply:
If you're looking for a nice api, might I suggest returning an `std::optional<glm::vec3>`

Reply:
good point I've swapped this to be pass by reference rather than the pointer.

regards `std::optional<glm::vec3` that's for the return type? what about if I want to use it to qualify if the ray actually hits? how would this work?

Reply:
Optional as a return type. Instead of returning false you would return an `std::nullopt` else you would return the vec3
https://en.cppreference.com/w/cpp/utility/optional

Reply:
Ok I've changed it somewhat to this and also returning an optional transform rather than just a position (I also want the normals for the camera path)